sound/soc/sof/intel/hda-dai.c

```diff
@@ -417,13 +426,23 @@ static int ipc4_hda_dai_trigger(struct snd_pcm_substream *substream,
 static int hda_dai_hw_free(struct snd_pcm_substream *substream,
 			   struct snd_soc_dai *dai)
 {
+	struct snd_soc_dapm_widget *w = snd_soc_dai_get_widget(dai, substream->stream);
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(dai->component);
+	const struct sof_ipc_tplg_ops *tplg_ops = sdev->ipc->ops->tplg;
+	struct snd_sof_widget *swidget = w->dobj.private;
+	struct snd_sof_dai_config_data data;
 	int ret;
 
 	ret = hda_link_dma_hw_free(substream, dai);
 	if (ret < 0)
 		return ret;
 
-	return hda_dai_hw_free_ipc(substream->stream, dai);
+	if (tplg_ops && tplg_ops->dai_config) {
+		data.dai_data = DMA_CHAN_INVALID;
+		return tplg_ops->dai_config(sdev, swidget, SOF_DAI_CONFIG_FLAGS_NONE, &data);
```

Comment:
I am really confused on this one. How is this equivalent between left and right?